### PR TITLE
Page/Tx-details

### PR DESCRIPTION
**Pull request by poomthiti**
## Describe your changes
Main features
1. Route and page setup for transaction details
2. useTxData hook
3. Top section
4. Side section (TxInfo)
5. Message handling switch cases and related components
6. Token card value display
7. Transaction not found ui

Minor changes
1. Minor refactoring and improvements of some component/hook

## Demo Link
https://osmosis-celatone-frontend-staging-git-pages-tx-page-alleslabs.vercel.app/testnet/tx/3D8EA04FE7E38E879D7B05B004E69B17E6D00B8F02806A8A89E5AE583D682370

### CODE REVIEW

File: src/lib/pages/tx-details/index.tsx [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import { TxHeader, TxInfo } from "./components";
+import { MessageSection } from "./components/MessageSection";
+
+const TxDetails = () => {
+  const router = useRouter();
+  const hashParam = getFirstQueryParam(router.query.txHash);
+
+  const { data: txData, isLoading } = useTxData(hashParam);
+
```
Comment (songwongtp) [inline src/lib/pages/tx-details/index.tsx:19]: add useEffect with AmpTrack

**Review by songwongtp — CHANGES_REQUESTED** [review]

**Comment by traviolus** [comment]
We have a comma-separated value for the Gas Used/Requested field in the Figma design na. Does not exist in this PR.

screenshot_2023-03-07_at_2 22 03_pm

**Comment by traviolus** [comment]
The `Title` field for /cosmos.gov.v1beta.MsgSubmitProposal is missing its value. See more here https://osmosis-celatone-frontend-staging-git-pages-tx-page-alleslabs.vercel.app/testnet/tx/186C2929D06F0D2921D4955A93C3C3929C2988685C62A0E3E4DA1104C80335DE

Screenshot 2023-03-07 at 2 41 25 PM

**Review by traviolus — CHANGES_REQUESTED** [review]
Please fix errors above

File: src/lib/types/currency/balance.ts [hunk 2]
```diff
@@ -3,10 +3,10 @@ import type { Option } from "lib/types";
 export interface Balance {
   amount: string;
   id: string;
-  name: Option<string>;
+  name?: Option<string>;
```
Comment (songwongtp) [inline src/lib/types/currency/balance.ts:6]: why both ? and Option

File: src/lib/types/currency/balance.ts [hunk 3]
```diff
@@ -3,10 +3,10 @@ import type { Option } from "lib/types";
...
   precision: number;
-  symbol: Option<string>;
-  type: string;
+  symbol?: Option<string>;
```
Comment (songwongtp) [inline src/lib/types/currency/balance.ts:8]: ditto

**Review by songwongtp — CHANGES_REQUESTED** [review]

File: src/lib/types/currency/balance.ts (hunk 2 above)
Comment (poomthiti) [reply src/lib/types/currency/balance.ts:6]: name and symbol field wont be included in Unsupported token case (those only present in Celatone API)

File: src/lib/pages/contract-details/components/InstantiateInfo.tsx [hunk 4]
```diff
@@ -148,7 +148,9 @@ export const InstantiateInfo = ({
 
       <LabelText
         label="Admin Address"
-        helperText1={instantiateInfo.admin ? adminType : undefined}
+        helperText1={
```
Comment (evilpeach) [inline src/lib/pages/contract-details/components/InstantiateInfo.tsx:151]: 1. It should look something like `helperText1={getAddressTypeText(adminType)}`.
2. Add the `validator_address` case on the `getAddressTypeText` function, as shown below.
```tsx
export const getAddressTypeText = (addressType: AddressReturnType) => {
  switch (addressType) {
    case "contract_address":
      return "(Contract Address)";
    case "user_address":
      return "(Wallet Address)";
    case "validator_address":
      return "(Validator Address)";
    case "invalid_address":
    default:
      return "(Invalid Address)";
  }
};
```

File: src/lib/utils/formatter/token.ts [hunk 5]
```diff
@@ -87,3 +87,8 @@ export const formatPrice = (value: USD<BigSource>): string => {
   }
   return `$${d6}`;
 };
+
+export const formatInteger = (n: BigSource): string => {
+  const formatter = formatDemimal({ decimalPoints: 0, delimiter: true });
+  return formatter(n, "FormatError");
```
Comment (evilpeach) [inline src/lib/utils/formatter/token.ts:93]: For me, It's strange to see the fallback value displayed as "FormatError".

File: src/lib/types/currency/balance.ts [hunk 6]
```diff
@@ -21,6 +21,7 @@ export interface AssetInfo {
   slugs: string[];
   symbol: string;
   type: string;
+  price: number;
```
Comment (evilpeach) [inline src/lib/types/currency/balance.ts:24]: Is it an optional value from API?

File: src/lib/services/tx.ts [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+  sequence: string;
+}
+
+interface ModeInfo {
+  single?: ModeInfoSingle | undefined;
+  multi?: ModeInfoMulti | undefined;
+}
+
+interface ModeInfoSingle {
+  mode: keyof typeof SignMode;
```
Comment (evilpeach) [inline src/lib/services/tx.ts:32]: Why not just use `SignMode`?

File: src/lib/components/ViewPermissionAddresses.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            <ExplorerLink
+              key={addr}
+              type={getAddressType(addr)}
+              value={addr}
+              canCopyWithHover
+            />
+          );
+        })}
+      {permissionAddresses.length > 1 && (
+        <Button
```
Comment (evilpeach) [inline src/lib/components/ViewPermissionAddresses.tsx:36]: image

I think the padding is incorrect.

File: src/lib/components/ViewPermissionAddresses.tsx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        permissionAddresses.map((addr) => {
+          return (
+            <ExplorerLink
+              key={addr}
+              type={getAddressType(addr)}
+              value={addr}
+              canCopyWithHover
+            />
+          );
+        })}
```
Comment (evilpeach) [inline src/lib/components/ViewPermissionAddresses.tsx:34]: remove return

File: src/lib/utils/tx/findAttr.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import type { Log } from "@cosmjs/stargate/build/logs";
```
Comment (evilpeach) [inline src/lib/utils/tx/findAttr.ts:1]:
```tsx
import { logs } from "@cosmjs/stargate";

import type { Option } from "lib/types";

export const findAttr = (
  log: Option<logs.Log>,
  eventType: string,
  attrKey: string
): Option<string> => {
  if (!log) return undefined;
  try {
    return logs.findAttribute([log], eventType, attrKey).value;
  } catch {
    return undefined;
  }
};
```

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/pages/faucet/index.tsx [hunk 11]
```diff
@@ -198,21 +196,24 @@ const Faucet = () => {
...
+              style={{ padding: "4px 12px" }}
+              onClick={() =>
+                window.open(
+                  `/${getNetworkByChainName(currentChainName)}/tx/${
+                    result.txHash
+                  }`,
+                  "_blank",
+                  "noopener,noreferrer"
+                )
+              }
```
Comment (evilpeach) [inline src/lib/pages/faucet/index.tsx:213]: Let's create util function for opening the new tab.

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/services/tx.ts [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+// Revisit Multisig Info
+interface ModeInfoMulti {
+  bitarray?: CompactBitArray;
+  modeInfos: ModeInfo[];
+}
+
+// ----------------------------------------
+// -----------------Tx---------------------
+// ----------------------------------------
+export type TypeUrl =
```
Comment (evilpeach) [inline src/lib/services/tx.ts:44]: move this to `/data/`

File: src/lib/pages/tx-details/index.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+  useEffect(() => {
+    if (router.isReady) AmpTrack(AmpEvent.TO_TRANSACTION_DETAIL);
+  }, [router.isReady]);
+
+  if (isLoading) return <Loading />;
```
Comment (evilpeach) [inline src/lib/pages/tx-details/index.tsx:26]: Before querying the data, I suggest we should offline validate the tx hash by checking the length.

File: src/lib/pages/tx-details/components/TxInfo.tsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    w: "250px",
+  },
+});
+
+export const TxInfo = ({ txData, ...flexProps }: TxInfoProps) => (
+  <Container {...flexProps}>
+    <LabelText label="Network">{txData.chainId}</LabelText>
+    <LabelText label="Block">
+      <ExplorerLink
+        value={txData.height.toString()}
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/TxInfo.tsx:26]: Is the height already a string type?

File: src/lib/pages/tx-details/components/TxInfo.tsx [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+        type="block_height"
+        canCopyWithHover
+      />
+    </LabelText>
+    <LabelText label="Transaction Fee">
+      {txData.formattedFee || (
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/TxInfo.tsx:32]: When the left hand is an optional value, I prefer to use '??' as a fallback value.

File: src/lib/pages/tx-details/components/TxInfo.tsx [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        <Text variant="body2" color="text.dark">
+          No Fee
+        </Text>
+      )}
+    </LabelText>
+    <LabelText label="Gas Used/Requested">
+      {`${formatInteger(txData.gas_used)}/${formatInteger(txData.gas_wanted)}`}
+    </LabelText>
+    <LabelText label="Memo">
+      {txData.tx.body.memo || (
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/TxInfo.tsx:42]: ditto

File: src/lib/pages/tx-details/components/TxHeader.tsx [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+  const chainId = useChainId();
+  const isTxFailed = Boolean(txData.code);
+  const openLcdPage = () => {
+    window.open(
+      `${CELATONE_API_ENDPOINT}/txs/${getChainApiPath(
+        currentChainName
+      )}/${chainId}/${txData.txhash}`,
+      "_blank",
+      "noopener,noreferrer"
+    );
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/TxHeader.tsx:31]: use util function

File: src/lib/services/txService.ts [hunk 18]
```diff
@@ -12,13 +16,48 @@ import {
...
   unwrapAll,
 } from "lib/utils";
 
+import type { TxResponse } from "./tx";
+import { queryTxData } from "./tx";
+
+export interface TxData extends TxResponse {
+  chainId: string;
+  formattedFee: Option<string>;
+}
```
Comment (evilpeach) [inline src/lib/services/txService.ts:33]: add `isTxFailed`

File: src/lib/pages/tx-details/components/MessageSection.tsx [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+
+interface MessageSectionProps {
+  txData: TxData;
+}
+
+export const MessageSection = ({ txData }: MessageSectionProps) => {
+  const isTxFailed = Boolean(txData.code);
+  const msgs = txData.tx.body.messages;
+  return (
+    <Flex direction="column" flex={0.98} gap={4}>
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/MessageSection.tsx:22]: 0.98?

File: src/lib/pages/tx-details/components/tx-message/index.tsx [hunk 20]
```diff
@@ -0,0 +1,29 @@
+import { Flex } from "@chakra-ui/react";
+import type { Log } from "@cosmjs/stargate/build/logs";
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/index.tsx:2]: fix import Log

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/pages/tx-details/components/tx-message/TxMsgExpand.tsx [hunk 21]
```diff
@@ -0,0 +1,311 @@
...
+
+  switch (type) {
+    case "/cosmwasm.wasm.v1.MsgStoreCode":
+      msgIcon = "upload";
+      content = (
+        <>
+          Upload Wasm
+          {log && (
+            <>
+              {" "}
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/TxMsgExpand.tsx:45]: If you already have gap attribute, should all of these `" "` be removed?

File: src/lib/pages/tx-details/components/tx-message/EventBox.tsx [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+
+export const EventBox = ({ event, msgIndex }: EventBoxProps) => {
+  const getAddressType = useGetAddressType();
+  const [expand, setExpand] = useState(true);
+
+  const receipts = event.attributes.map<TxReceipt>(({ key, value }) => {
+    const addrType = getAddressType(value);
+    let valueComponent: ReactNode;
+
+    switch (true) {
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/EventBox.tsx:27]: use `if`.

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/index.tsx [hunk 23]
```diff
@@ -0,0 +1,1957 @@
...
+            <Flex gap={3} align="flex-start">
+              Size:{" "}
+              {big(Buffer.from(details.wasm_byte_code).byteLength)
+                .div(1024)
+                .toFixed(1)}{" "}
+              KB
+              <CopyButton
+                value={details.wasm_byte_code}
+                variant="ghost-primary"
+                tooltipBgColor="lilac.darker"
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/index.tsx:94]: It's no need to pass `tooltipBgColor`. just use `honeydew.darker`

File: src/lib/components/copy/CopyButton.tsx [hunk 24]
```diff
@@ -1,43 +1,54 @@
...
 
-interface CopyButtonProps {
+interface CopyButtonProps extends ButtonProps {
   isDisable?: boolean;
   value: string;
-  size?: ButtonProps["size"];
   copyLabel?: string;
+  hasIcon?: boolean;
+  buttonText?: string;
+  tooltipBgColor?: TooltipProps["bgColor"];
```
Comment (evilpeach) [inline src/lib/components/copy/CopyButton.tsx:15]: remove

File: src/lib/pages/tx-details/components/tx-message/EventBox.tsx [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+          <ExplorerLink
+            type={addrType}
+            value={value}
+            canCopyWithHover
+            textFormat="normal"
+            maxWidth="full"
+          />
+        );
+        break;
+      case key === "code_id":
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/EventBox.tsx:39]: Merge `code_id` and `proposal_id` case.

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+export const getCommonReceiptHtml = <T extends HtmlType>({
+  type,
+  value,
+  linkType = "invalid_address",
+  fallback,
+}: CommonReceiptHtmlArgs<T, T extends "json" ? object : string>) => {
+  if (!value)
+    return (
+      <Text variant="body2" color="text.dark">
+        {fallback}
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:38]: default fallback value?

File: src/lib/pages/tx-details/components/TxInfo.tsx (hunk 16 above)
Comment (poomthiti) [reply src/lib/pages/tx-details/components/TxInfo.tsx:42]: not an option but mostly is empty string

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+  );
+};
+
+export const getCoinComponent = (
+  amount: Coin | Coin[],
+  assetInfos: Option<{ [key: string]: AssetInfo }>
+) => <CoinComponent amount={amount} assetInfos={assetInfos} />;
+
+export const getGenericValueEntry = (
+  entry: [string, string | object],
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:67]: replace `entry` by `[title, value]`

File: src/lib/components/copy/CopyButton.tsx [hunk 28]
```diff
@@ -1,43 +1,53 @@
...
 
 export const CopyButton = ({
   isDisable,
   value,
   size = "sm",
   copyLabel,
+  hasIcon = true,
+  variant = "outline-info",
+  buttonText = "Copy",
+  tooltipBgColor,
```
Comment (evilpeach) [inline src/lib/components/copy/CopyButton.tsx:26]: remove

File: src/lib/components/copy/CopyButton.tsx [hunk 29]
```diff
@@ -1,43 +1,53 @@
...
 
-interface CopyButtonProps {
+interface CopyButtonProps extends ButtonProps {
   isDisable?: boolean;
   value: string;
-  size?: ButtonProps["size"];
   copyLabel?: string;
+  hasIcon?: boolean;
+  buttonText?: string;
+  tooltipBgColor?: TooltipProps["bgColor"];
```
Comment (evilpeach) [inline src/lib/components/copy/CopyButton.tsx:15]: remove?

File: src/lib/types/currency/balance.ts [hunk 30]
```diff
@@ -21,6 +21,7 @@ export interface AssetInfo {
   slugs: string[];
   symbol: string;
   type: string;
+  price: Option<number>;
```
Comment (evilpeach) [inline src/lib/types/currency/balance.ts:24]: `price?: number` or `price: Option<number>`?

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by tansawit — CHANGES_REQUESTED** [review]
incorrect price. prices needs to be queried from `<CELATONE_API>/assets/<CHAIN>/<NETWORK>/prices`

CleanShot 2566-03-22 at 16 11 22@2x

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+  linkType?: LinkType;
+  fallback?: string;
+}
+
+// Util Functions
+export const getCommonReceiptHtml = <T extends HtmlType>({
+  type,
+  value,
+  linkType = "invalid_address",
+  fallback = "-",
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:33]: N/A

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+}: CommonReceiptHtmlArgs<T, T extends "json" ? object : string>) => {
+  if (!value)
+    return (
+      <Text variant="body2" color="text.dark">
+        {fallback}
+      </Text>
+    );
+
+  // TODO: Find a solution, TS doesn't know that type === 'json' would make typeof value === 'object
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:42]: 'object'

**Review by evilpeach — CHANGES_REQUESTED** [review]

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 33]
```diff
@@ -0,0 +1,167 @@
...
+}
+
+// Util Functions
+export const getCommonReceiptHtml = <T extends HtmlType>({
+  type,
+  value,
+  linkType = "invalid_address",
+  fallback = "N/A",
+}: CommonReceiptHtmlArgs<T, T extends "json" ? object : string>) => {
+  if (value === undefined)
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:35]: I recommend that you do the following:
```tsx
export const getCommonReceiptHtml = <T extends HtmlType>({
  type,
  value,
  linkType = "invalid_address",
  fallback,
}: CommonReceiptHtmlArgs<T, T extends "json" ? object : string>) => {
  switch (true) {
    case value === null:
      return (
        <Text variant="body2" color="pebble.600">
          null
        </Text>
      );
    case value === undefined && fallback:
      return (
        <Text variant="body2" color="text.dark">
          {fallback}
        </Text>
      );
    case value === undefined:
      return (
        <Text variant="body2" color="warning.dark">
          Data not found
        </Text>
      );
    // TODO: Find a solution, TS doesn't know that type === "json" would make typeof value === "object"
    case type === "json" || typeof value === "object":
      return (
        <JsonReadOnly
          text={JSON.stringify(value, null, 2)}
          canCopy
          fullWidth
          isExpandable
        />
      );
    default:
      return (
        <ExplorerLink
          type={linkType}
          value={value as string}
          showCopyOnHover
          textFormat="normal"
          maxWidth="full"
        />
      );
  }
};```

File: src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx [hunk 34]
```diff
@@ -0,0 +1,167 @@
...
+  linkType?: LinkType;
+  fallback?: string;
+}
+
+// Util Functions
+export const getCommonReceiptHtml = <T extends HtmlType>({
+  type,
+  value,
+  linkType = "invalid_address",
+  fallback = "N/A",
```
Comment (evilpeach) [inline src/lib/pages/tx-details/components/tx-message/msg-receipts/renderUtils.tsx:33]: There is "Data not found" fallback case. It's no need to use `N/A`.

**Review by evilpeach — APPROVED** [review]
lgtm

**Review by songwongtp — APPROVED** [review]

**Review by bkioshn — APPROVED** [review]

**Review by tansawit — APPROVED** [review]

**Pull request merged**